src/Compilers/CSharp/Portable/Binder/PatternSwitchBinder.cs

```diff
@@ -138,15 +135,41 @@ internal override void BindPatternSwitchLabelForInference(CasePatternSwitchLabel
             Debug.Assert(sectionBinder != null);
             var labelsByNode = LabelsByNode;
 
+            bool? inputMatchesType(TypeSymbol patternType)
+            {
+                // use-site diagnostics will have been reported previously.
+                HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+                return ExpressionOfTypeMatchesPatternType(Conversions, SwitchGoverningType, patternType, ref useSiteDiagnostics, out _, SwitchGoverningExpression.ConstantValue, true);
+            }
+
             foreach (var labelSyntax in node.Labels)
             {
                 LabelSymbol label = labelsByNode[labelSyntax];
-                BoundPatternSwitchLabel boundLabel = BindPatternSwitchSectionLabel(sectionBinder, boundSwitchExpression, labelSyntax, label, ref defaultLabel, diagnostics);
-                bool valueMatched; // true if we find an unconditional constant label that matches the input constant's value
-                bool isReachable = subsumption.AddLabel(boundLabel, diagnostics, out valueMatched);
-                boundLabel = boundLabel.Update(boundLabel.Label, boundLabel.Pattern, boundLabel.Guard, isReachable && !someValueMatched);
-                someValueMatched |= valueMatched;
+                BoundPatternSwitchLabel boundLabel = BindPatternSwitchSectionLabel(sectionBinder, labelSyntax, label, ref defaultLabel, diagnostics);
+                bool isNotSubsumed = subsumption.AddLabel(boundLabel, diagnostics);
+                bool guardAlwaysSatisfied = boundLabel.Guard == null || boundLabel.Guard.ConstantValue == ConstantValue.True;
+
+                // patternMatches is true if the input expression is unconditionally matched by the pattern, false if it never matches, null otherwise.
+                // While subsumption would produce an error for an unreachable pattern based on the input's type, this is used for reachability (warnings),
+                // and takes the input value into account.
+                bool? patternMatches =
+                    labelSyntax.Kind() == SyntaxKind.DefaultSwitchLabel ? null :
+                    // wildcard pattern matches anything
+                    boundLabel.Pattern.Kind == BoundKind.WildcardPattern ? true :
+                    // `var x` matches anything
+                    // `Type x` matches anything of a subtype of `Type` except null
+                    boundLabel.Pattern is BoundDeclarationPattern d ? (d.IsVar ? true : inputMatchesType(d.DeclaredType.Type)) :
+                    // `case 2` matches the input `2`
+                    boundLabel.Pattern is BoundConstantPattern p ? SwitchGoverningExpression.ConstantValue?.Equals(p.ConstantValue) :
+                    null;
```

Comment:
>null; [](start = 20, length = 5)

I think this code would be easier to read and maintain if it was written as a sequence of ```if-else if-else``` statements